libraries/node-core-library/src/Executable.ts

```diff
@@ -0,0 +1,419 @@
+   *
+   * @remarks
+   * If the name has a relative path, it will be resolved relative to the current working directory.
+   * If the name has no path, then the shell PATH will be searched.  If the name is missing a Windows
+   * file extension, it may be appended (based on the PATHEXT environment variable).
+   *
+   * @param name - the name of the executable, which may be missing the path or file extension
+   * @param options - optional other parameters
+   * @returns the absolute path of the executable, or undefined if it was not found
+   */
+  public static tryResolve(name: string, options?: IExecutableResolveOptions): string | undefined {
+    return Executable._tryResolve(name, options || { }, Executable._getExecutableContext(options));
+  }
+
+  private static _tryResolve(name: string, options: IExecutableResolveOptions,
+    context: IExecutableContext): string | undefined {
+
+    const hasPathSeparators: boolean = name.indexOf('/') >= 0
+      || (os.platform() === 'win32' && name.indexOf('\\') >= 0);
+
+    const pathsToSearch: string[] = [];
+
+    // Are there any path separators?
+    if (hasPathSeparators) {
+      // If so, then only search the resolved path
+      pathsToSearch.push(path.resolve(context.currentWorkingDirectory, name));
+    } else {
+      // Otherwise if it's a bare name, then try everything in the shell PATH
+      pathsToSearch.push(...Executable._getSearchFolders(context));
+    }
+
+    for (const pathToSearch of pathsToSearch) {
+      const resolvedPath: string = path.join(pathToSearch, name);
+
+      if (Executable._canExecute(resolvedPath, context)) {
+        return resolvedPath;
+      }
+
+      // Try the default file extensions
+      for (const shellExtension of context.windowsExecutableExtensions) {
+        const resolvedNameWithExtension: string = resolvedPath + shellExtension;
+
+        if (Executable._canExecute(resolvedNameWithExtension, context)) {
+          return resolvedNameWithExtension;
+        }
+      }
+    }
+
+    return undefined;
+  }
+
+  /**
+   * This is used when searching the shell PATH for an executable, to determine
+   * whether a match should be skipped or not.  If it returns true, this does not
+   * guarantee that the file can be successfully executed.
+   */
+  private static _canExecute(filePath: string, context: IExecutableContext): boolean {
+    if (!FileSystem.exists(filePath)) {
+      return false;
+    }
+
+    const fileExtension: string = path.extname(filePath);
+
+    if (os.platform() === 'win32') {
+      // Does the file have an executable file extension?
+      let matchFound: boolean = false;
+      for (const executableExtension of context.windowsExecutableExtensions) {
+        if (fileExtension.localeCompare(executableExtension) === 0) {
+          matchFound = true;
+          break;
+        }
+      }
+
+      if (fileExtension.localeCompare('.js') === 0) {
+        // The .js extension is special and always allowed since this is NodeJS.
```

Comment:
>The .js extension is special and always allowed since this is NodeJS. [](start = 11, length = 69)

I think this should be specific to Rushell.

Reply:
Agreed

---
In reply to: [208422863](https://github.com/Microsoft/web-build-tools/pull/764#discussion_r208422863) [](ancestors = 208422863)